session/nontransactional.go

```
@@ -0,0 +1,492 @@
+				Op: opcode.LogicOr,
+				L:  leCondition,
+				R:  isNullCondition,
+			}
+		}
+	} else {
+		// a normal between condition: `where x between start and end`
+		left := &driver.ValueExpr{}
+		left.Type = tp
+		left.Datum = job.start
+		right := &driver.ValueExpr{}
+		right.Type = tp
+		right.Datum = job.end
+		whereCondition = &ast.BetweenExpr{
+			Expr: &ast.ColumnNameExpr{
+				Name:  stmt.ShardColumn,
+				Refer: refer,
+			},
+			Left:  left,
+			Right: right,
+			Not:   false,
+		}
+	}
+
+	if originalCondition == nil {
+		stmt.DeleteStmt.Where = whereCondition
+	} else {
+		stmt.DeleteStmt.Where = &ast.BinaryOperationExpr{
+			Op: opcode.LogicAnd,
+			L:  whereCondition,
+			R:  originalCondition,
+		}
+	}
+	var sb strings.Builder
+	err := stmt.DeleteStmt.Restore(format.NewRestoreCtx(format.DefaultRestoreFlags|
+		format.RestoreNameBackQuotes|
+		format.RestoreSpacesAroundBinaryOperation|
+		format.RestoreBracketAroundBinaryOperation|
+		format.RestoreStringWithoutCharset, &sb))
+	if err != nil {
+		job.err = err
+		return ""
+	}
+	deleteSQL := sb.String()
+
+	if stmt.DryRun == ast.DryRunSplitDml {
+		return deleteSQL
+	}
+
+	job.sql = deleteSQL
+	stmt.DeleteStmt.SetText(nil, fmt.Sprintf("/* job %v/%v */ %s", job.jobID, totalJobCount, deleteSQL))
+	rs, err := se.ExecuteStmt(context.TODO(), stmt.DeleteStmt)
+
+	// collect errors
+	failpoint.Inject("splitDeleteError", func(_ failpoint.Value) {
+		err = errors.New("injected split delete error")
+	})
+	if err != nil {
+		errStr := fmt.Sprintf("Non-transactional delete SQL failed, sql: %s, error: %s, jobID: %d, jobSize: %d. ",
+			deleteSQL, err.Error(), job.jobID, job.jobSize)
+		logutil.Logger(ctx).Error(errStr)
+		job.err = err
+	} else {
+		logutil.Logger(ctx).Info("Non-transactional delete SQL finished successfully", zap.Int("jobID", job.jobID),
+			zap.Int("jobSize", job.jobSize), zap.String("deleteSQL", deleteSQL))
+	}
+	if rs != nil {
+		rs.Close()
+	}
+	return ""
```

Better to return the processing error and keep it consistent.